src/runtime_strings.js

```diff
@@ -16,24 +16,27 @@ var UTF8Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf8') :
  * @param {number=} maxBytesToRead
  * @return {string}
  */
-function UTF8ArrayToString(u8Array, idx, maxBytesToRead) {
+function UTF8ArrayToString(heap, idx, maxBytesToRead) {
+#if CAN_ADDRESS_2GB
+  idx >>>= 0;
```

Comment:
Likewise, this function is never called directly from wasm, but from a Number... it seems to be that the shift should have occurred already earlier?

Reply:
It's true that optimally the shift would occur in an earlier place. But imagine we have user code like
```
var x = foo();
x = x + 10 | 0;
var y = UTF8ArrayToString(HEAP8, x);
```
Given user-written JS code, I don't think we can assume that they will properly keep the value unsigned. I think the safe thing is to treat the small number of pointer-using functions (like UTF8ArrayToString) as receiving input with uncertain sign, so we fix up the sign, just like at the beginning of an asm.js function.

Basically, my logic is similar to LLVM or asm.js: pointers can be passed around and stored with incorrect sign, but when we do an operation where the sign matters, we make sure to set the sign correctly. In these functions it is easy to do that by just unsigning the input (but a few cases do require more unsign operations in the middle, replacing `| 0`s etc.).

Reply:
> Given user-written JS code, I don't think we can assume that they will properly keep the value unsigned.

I think we can, and we should? The code example you posted would be invalid for user code to have in the first place under the 4GB address space mode.

> I think the safe thing is to treat the small number of pointer-using functions (like UTF8ArrayToString) as receiving input with uncertain sign, so we fix up the sign, just like at the beginning of an asm.js function.

>  but when we do an operation where the sign matters, we make sure to set the sign correctly.

The danger here might be that if we do the shifts in utility functions to patch up "after the fact", then the shift only applies to those utility functions, and users might still be doing other operations to the data in their own code, and those operations would be then incorrect. Then they'll still need to fix up their own code, and we result in duplicate/redundant shifts.

In other words, it feels like these shifts are only a partial fix "in the wrong place", whereas the "proper" fix would be to have the shifts right when the call is made.

To me it seems that like C string -> JS string marshalling, we should have a rule/convention for marshalling uint32->Number back from wasm that users need to adhere to, either manually or by using a `{{{ callWasm(...) }}}` type of construct.

I think the only exception to this is perhaps the marshalling related functions, UTF8ToString() etc. since it is more convenient to call `UTF8ToString(_someWasmFuncThatReturnsAPtr())` rather than always needing to do `UTF8ToString(_someWasmFuncThatReturnsAPtr()>>>0)` - but that exception seems to be appropriate specifically for functions that deal with type wasm-JS type marshalling.

But for other library code that accesses the heap with pointers - it seems that there are boatloads of library functions that do heap accesses, vs the number of functions that call back from wasm; doing the shifts in the wasm-js boundary would result in fewer shifts overall vs annotating all heap accesses in all library functions?

Reply:
In general for heap accesses we have https://github.com/emscripten-core/emscripten/pull/10601 that will instrument them automatically. It also only operates when 2GB+ is relevant, so we don't increase code size in <2GB builds (as `>>` is smaller than `>>>`).

So the remaining risks are when JS does manual computation of pointers and then uses them directly, like say using the pointer as the key of a hash map, or comparing between two pointers.

I agree this PR might "paper over" problems in user code. But I really think getting this right in user code is a high bar - this is a completely new requirement for users to think about, that only happens in 2GB+ heaps, and will likely be a cause of bugs for them. We should document this a lot more, but I also think `>>>=`s in key places in our API, where we receive pointers from users *and* then proceed to use them in ways where the sign matters (rare combination), we should fix up the sign.